crates/fs/public/directory.rs

```diff
@@ -538,10 +537,10 @@ impl PublicDirectory {
     }
 
     /// Gets the iterator for walking the history of a directory node.
-    pub fn get_history<'a, B: BlockStore>(
+    pub fn get_history<B: BlockStore>(
         self: Rc<Self>,
-        store: &'a B,
-    ) -> impl Stream<Item = Result<Cid>> + 'a {
+        store: &B,
+    ) -> impl Stream<Item = Result<Cid>> + '_ {
```

Comment:
Huh interesting. So that's the same as saying
```rust
pub fn get_history<'a, B: BlockStore>(
    self: Rc<Self>,
    store: &B,
) -> impl Stream<Item = Result<Cid>> + 'a
```
right?

I'm not quite sure what exactly is happening here. Why do we have to annotate this lifetime, it's not like we return something borrowed, right? The caller will own the stream, no?

(This is me learning rust again, it's not about changing our code here :D )

Reply:
The `try_stream!` macro creates a Stream object that gets returned in this function. And Streams, like Futures and Closures, can capture references from outer scope. In this case the Stream is borrowing the `store` object and the compiler has to make sure `store` lifetime exceeds that of the returned Stream object, hence the annotation on the return type.

Because writing lifetimes can be unwieldy sometimes, Rust introduced lifetime elision rules, which are some of the common cases where the compiler is able to infer a lifetime so you don't have to be explicit about it. You can read up on the rules for elision here. https://doc.rust-lang.org/nomicon/lifetime-elision.html

In this case, there is only one input lifetime (`store: &B`), which automatically propagates to all output lifetimes, following the second rule in that book.

Even though we need to specify that there is a lifetime on the output, we don't need to specify which lifetime.